src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,326 @@
+torrentAboutToBeRemoved);
+
+    populate();
+}
+
+TagFilterModel::~TagFilterModel()
+{
+    foreach (const TagModelItem *item, m_tagItems.values())
+        delete item;
+}
+
+bool TagFilterModel::isSpecialItem(const QModelIndex &index)
+{
+    // the first two items are special items: 'All' and 'Untagged'
+    return (!index.parent().isValid() && (index.row() <= 1));
+}
+
+int TagFilterModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant TagFilterModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid() || index.column() != 0)
+        return QVariant();
+
+    auto item = static_cast<TagModelItem *>(index.internalPointer());
+
+    if (role == Qt::DecorationRole)
+        return GuiIconProvider::instance()->getIcon("inode-directory");
+    if (role == Qt::DisplayRole)
+        return QString(QStringLiteral("%1 (%2)"))
+               .arg(item->displayName()).arg(item->torrentsCount());
+    if (role == Qt::UserRole)
+        return item->torrentsCount();
+    return QVariant();
+}
+
+Qt::ItemFlags TagFilterModel::flags(const QModelIndex &index) const
+{
+    if (!index.isValid())
+        return 0;
+    return Qt::ItemIsEnabled | Qt::ItemIsSelectable;
+}
+
+QVariant TagFilterModel::headerData(int section, Qt::Orientation orientation, int role) const
+{
+    if ((orientation == Qt::Horizontal) && (role == Qt::DisplayRole))
+        if (section == 0)
+            return tr("Tags");
+    return QVariant();
+}
+
+QModelIndex TagFilterModel::index(int row, int column, const QModelIndex &parent) const
+{
+    if (parent.isValid() || (row >= m_tagRows.size()) || (column != 0))
+        return QModelIndex();
+    return createIndex(row, column, m_tagItems[m_tagRows.at(row)]);
+}
+
+QModelIndex TagFilterModel::parent(const QModelIndex &) const
+{
+    return QModelIndex();  // Tags are single-level.
+}
+
+int TagFilterModel::rowCount(const QModelIndex &parent) const
+{
+    if (!parent.isValid())
+        return m_tagRows.count();
+    return 0;
+}
+
+QModelIndex TagFilterModel::index(const QString &tag) const
+{
+    return index(findItem(tag));
+}
+
+QString TagFilterModel::tag(const QModelIndex &index) const
+{
+    if (!index.isValid()) return QString();
+    return static_cast<TagModelItem *>(index.internalPointer())->tag();
+}
+
+QModelIndex TagFilterModel::index(const TagModelItem *item) const
+{
+    if (!item)
+        return QModelIndex();
+    const int row = m_tagRows.indexOf(item->tag());
+    if (row < 0)
+        return QModelIndex();
+    return index(row, 0, QModelIndex());
+}
+
+void TagFilterModel::tagAdded(const QString &tag)
+{
+    const int row = m_tagRows.count();
+    beginInsertRows(QModelIndex(), row, row);
+    addItemToModel(new TagModelItem(tag, tag));
+    endInsertRows();
+}
+
+void TagFilterModel::tagRemoved(const QString &tag)
+{
+    TagModelItem *item = findItem(tag);
+    if (item) {
+        QModelIndex i = index(item);
```

Comment:
Yes, this is as intended. They are searching for different items. The first one searches for the special "Untagged" entry, the second one searches for the tag that was just added.